pyro/infer/autoguide/effect.py

```diff
@@ -0,0 +1,409 @@
+        that don't have upstream sites, the guide is representing/learning deviation from the prior.
+    """
+
+    def __init__(
+        self,
+        model: Callable,
+        *,
+        init_loc_fn: Callable = init_to_mean(fallback=init_to_feasible),
+        init_scale: float = 0.1,
+        init_weight: float = 1.0,
+        hierarchical_sites: Optional[list] = None,
+    ):
+        if not isinstance(init_scale, float) or not (init_scale > 0):
+            raise ValueError("Expected init_scale > 0. but got {}".format(init_scale))
+        super().__init__(model)
+        self.init_loc_fn = init_loc_fn
+        self._init_scale = init_scale
+        self._init_weight = init_weight
+        self._hierarchical_sites = hierarchical_sites
+        self._computing_median = False
+
+    def get_posterior(
+        self,
+        name: str,
+        prior: Distribution,
+        upstream_values: Dict[str, torch.Tensor],
+    ) -> Union[Distribution, torch.Tensor]:
+        if self._computing_median:
+            return self._get_posterior_median(name, prior)
+
+        with helpful_support_errors({"name": name, "fn": prior}):
+            transform = biject_to(prior.support)
+        if (self._hierarchical_sites is None) or (name in self._hierarchical_sites):
+            # If hierarchical_sites not specified all sites are assumed to be hierarchical
+            loc, scale, weight = self._get_params(name, prior)
+            loc = loc + transform.inv(prior.mean) * weight
+            posterior = dist.TransformedDistribution(
+                dist.Normal(loc, scale).to_event(transform.domain.event_dim),
+                transform.with_cache(),
+            )
+            return posterior
+        else:
+            # Fall back to mean field when hierarchical_sites list is not empty and site not in the list.
+            return super().get_posterior(name, prior, upstream_values)
+
+    def _get_params(self, name: str, prior: Distribution):
+        try:
+            loc = deep_getattr(self.locs, name)
+            scale = deep_getattr(self.scales, name)
+            if (self._hierarchical_sites is None) or (name in self._hierarchical_sites):
+                weight = deep_getattr(self.weights, name)
+                return loc, scale, weight
+            else:
+                return loc, scale
+        except AttributeError:
+            pass
+
+        # Initialize.
+        with poutine.block(), torch.no_grad():
+            transform = biject_to(prior.support)
+            event_dim = transform.domain.event_dim
+            constrained = self.init_loc_fn({"name": name, "fn": prior}).detach()
+            unconstrained = transform.inv(constrained)
+            init_loc = self._remove_outer_plates(unconstrained, event_dim)
+            init_scale = torch.full_like(init_loc, self._init_scale)
+            # weight is a single value parameter
+            init_weight = torch.full((), self._init_weight)
```

Comment:
Re: plates, can you say more about your single-value design decision?  I can see why you might want to exclude `event_shape` from this, but [Gorinova et al. (2019)](https://arxiv.org/abs/1906.03028) observe that element-wise weights empirically outperformed single global weights in a similar context.  I can also see how you might want to amortize over a particular plate so as to support subsampling, but for that I think we could provide a more flexible mechanism such as a user specified `amortized_plates` kwarg to the `AutoMessenger.__init__()` (and then use the logic in `._remove_outer_plates()`), or even automatically amortize over any plate that has been subsampled.

Reply:
My intuition was that element-wise weights are over-parameterised because you can get the same loc by either adding `loc_independent` to `prior.mean` or by weighting `prior.mean`:

```python
loc = loc_independent + transform.inv(prior.mean) * weight
20 = 0 + 10 * 2
20 = 10 + 10 * 1
```

WDYT about adding several options that can be chosen by changing class attributes?
 1. Scalar weight
 2. Element-wise weights (same shape as init_loc). In this case, `weights` should be amortised exactly the same way as `loc` and `scales`.
 3. Exclude `event_shape`. I don't know how to do this.